lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -61,9 +61,24 @@ func (d *zfs) CreateVolume(vol Volume, filler *VolumeFiller, op *operations.Oper
 		if exists {
 			canRestore := true
 
-			// For block volumes check if the cached image volume is larger than the current pool volume.size
-			// setting (if so we won't be able to resize the snapshot to that the smaller size later).
-			if vol.contentType == ContentTypeBlock {
+			zfsVolType, err := d.getDatasetProperty(d.dataset(vol, true), "type")
+			if err != nil {
+				return err
+			}
+
+			if zfsVolType == "filesystem" && d.isBlockBacking(vol) || zfsVolType == "volume" && !d.isBlockBacking(vol) {
+				d.logger.Debug("Renaming deleted cached image volume so that regeneration is used", logger.Ctx{"fingerprint": vol.Name()})
+				randomVol := NewVolume(d, d.name, vol.volType, vol.contentType, fmt.Sprintf("%s_%s", vol.name, uuid.New()), vol.config, vol.poolConfig)
```

Comment:
I think we could do with a helper function called `randomVolumeName(vol Volume) string` that encompasses this logic:

```
fmt.Sprintf("%s_%s", vol.name, uuid.New())
```

That way it can be used here, and on line 110 to avoid duplicating the logic and making it easier to see where the same concepts are used.